__author__ = 'Daniel Wiegand - Wyss Institute - Harvard University'
__name__ = 'Preparation of Oligonucleotide Library for Multiplex DNA Assembly'

# ~~~~~~~~Version History~~~~~~~~~~
# 9.6.2016 - Added forward primer sequence at 15-bp for pre-subpooling enrichment & conversion into dsDNA
# 9.6.2016 - Added randomized barcode selection for plasmid amplification; check if it exists in the library
# 9.8.2016 - Removed randomized barcode selection, added hamming distance function to check orthogonality
# 9.14.2016 - Updated modifier section and altered design to include 121-nt payloads, 10-state and then 4-state
# 9.14.2016 - Updated some constant variable and cleaned code
# 9.21.2016 - Finalized fragment generation and decoded a few bugs in the script ---- v2
# 11.06.2016 - Fixed bug that added the wrong primer sequences to the final list; resynthesize library.
# 01.10.2017 - Moved temperature analysis module for subpooling primers from Hamming Distance module for efficiency.


import collections
import math
import primer3

from Bio.Seq import Seq
from Bio import SeqIO
from Bio.SeqRecord import SeqRecord
from Bio.Alphabet import IUPAC

def averageLen(list):
    lengths = [len(i) for i in list]
    return 0 if len(lengths) == 0 else (float(sum(lengths)) / len(lengths))

def hammingDistance(sequence, barcode):
    a = 0
    b = len(barcode)
    rc = Seq(barcode).reverse_complement()
    xx = []
    yy = []

    while b < len(sequence):
        xx.append(sum(i != k for i, k in zip(sequence[a:b], barcode)))
        yy.append(sum(i != k for i, k in zip(sequence[a:b], rc)))
        a += 1
        b += 1
    return 1 if min(xx) <= 9 or min(yy) <= 9 else 0

def BuildLibrary(library, barcodes, frag, oligo):
    # Setup primer modification, reverse pool primer has initial restriction site
    fpool = 'ATTGCGTTTCGCCAT'
    rpool = 'CactgcCCTCGGTTA'
    #Nicking site present, only the first of the sequences get this
    fassembly = 'cgtagtgttctcgctgtttg'
    rassembly = 'agaacccaacgtatgaccaa'
    dummy = 'TGGGGCTGTGCCCAGTTTCTGTTCCGGCCACTATGGGCTTTATGTGGTCACTGCTTGGCTGGGTCATCTTTGCTGCTCCATTGTCC' \
            'TTGGCCCTTCAGTAGAACCTTGTCCCACTAAGACCTGTGATCACAGGGAGTTGGATGTCACCTAGGGTGGTCCCTGCATACAAATC' \
            'TCCTTGCGGTATCAAGAGACAAATTTTCAGACCTGTCTAGGTCTTGCCTTCCTCCCAGGGCTTTTTCCTCAATTGTATTTTCAATT' \
            'TTTCTCCAATCTTTTTAAAGGAACCAGATTGTGACATTTGCA'
    frestriction = 'gCAATg' #BsrDI
    rrestriction = 'cATTGc' #BsrDI
    payload = oligo + 20

    thermoAnalysis = []

    # Parse sequences from fasta
    seq_library = SeqIO.parse(open(library), 'fasta')
    barcode_library = SeqIO.parse(open(barcodes), 'fasta')
    plasmid_map = SeqIO.parse(open('plasmid.fas'), 'fasta')
    s_bar = SeqIO.parse(open('screened.fas'), 'fasta')
    r_bar = SeqIO.parse(open('screened.fas'), 'fasta')

    # Setup ordered dictionaries for sequence and description
    my_assembly = collections.OrderedDict()
    my_barcodes = collections.OrderedDict()
    screened_bar = collections.OrderedDict()
    screened_sub = collections.OrderedDict()
    bar_orderlist = collections.OrderedDict()
    my_homology = []

    # Setup variable libraries for code
    lib_export = []
    bar_export = []
    order_barcodes = []
    block = []
    export = []

    # Process plasmid homology addition
    for plasmid in plasmid_map:
        my_homology.append(str(plasmid.seq))

    # Parse barcode list generated by SK
    for barcode in barcode_library:
        id, sequence = barcode.description, str(barcode.seq)
        my_barcodes[id] = sequence.lower()

    # Process sequences -- add homology here
    for seq in seq_library:
        id, sequence = seq.description, str(seq.seq)
        my_assembly[id] = sequence.lower()
        my_assembly[id] = my_homology[0] + sequence.lower() + my_homology[1]

    # Process screened barcodes for homology
    for b in s_bar:
        id, sequence = b.description, str(b.seq)
        screened_bar[id] = sequence.lower()

    # Process screened barcodes with restriction site for subpool
    for pool in r_bar:
        id, sequence = pool.description, str(pool.seq)
        screened_sub[id] = sequence.lower()

    # Generate list of barcodes based on library and store for future use.
    # Fix this to run faster and greater stringency with primer3
    if not screened_bar:

        print('Screening library.')

        i = 0
        while len(screened_bar) != (len(my_assembly)*9):
            ran_bar = 'TCGCCAT' + ((list(my_barcodes.items())[i])[1])[:13]

            h = []
            for k in my_assembly.items():
                h.append(hammingDistance(k[1], ran_bar))

            ppp = ((primer3.calcTm(ran_bar.upper(),10,5.0,0.5,500)))
            print(int(ppp))

            if 1 in h:
                print('Barcode ' + ran_bar + ' is incompatible with the library.')
                i += 1
            elif int(ppp) > 62 or int(ppp) < 58:
                print('Barcode ' + ran_bar + ' has an incompatible melting temperature')
                i += 1
            else:
                screened_bar[(list(my_barcodes.items())[i])[0]] = ran_bar
                print('Barcode ' + ran_bar + ' added. Total is now ' + (str(len(screened_bar))) + ' out of ' + str(len(my_assembly)*9))
                i += 1

        for j in range(0, len(screened_bar)):
            p = SeqRecord(Seq((list(screened_bar.items())[j])[1], IUPAC), id='DW_HLA-A-' + str(j), description='')
            bar_export.append(p)

        output_handle = open("screened.fas", 'w')
        SeqIO.write(bar_export, output_handle, "fasta")
        output_handle.close()
        print('Exported new list of screened barcodes.')

    # Add homology barcode and remaining sequence
    for v in range(0, len(my_assembly.items())):
        seq = (list(my_assembly.items())[v])[1]
        id = (list(my_assembly.items())[v])[0]
        bar = (list(screened_bar.items())[v])[1]
        my_assembly[id] = seq.lower()

        #my_assembly[id] = seq.lower() + bar + my_homology[2].lower()

        bar_orderlist[id] = bar

    # Prepare sequences for secondary overlap assembly with 40-bp overlaps
    for id, sequence in my_assembly.items():
        avgFrag = math.ceil(len(sequence) / frag)
        seqLength = len(sequence)

        divide = [sequence[i:i + avgFrag] for i in range(0, seqLength, avgFrag)]

        # Add overlaps for secondary assembly
        j = [divide[0]]
        for i in range(1, len(divide)):
            SeqA = divide[i - 1]
            SeqB = divide[i]
            j.append(SeqA[-40:].upper() + SeqB)

        for a in j:
            slice = [a[i:i + oligo] for i in range(0, len(a), oligo)]

            k = [fassembly + slice[0]]
            for p in range(1, len(slice)):
                SeqC = slice[p - 1]
                SeqD = slice[p]
                k.append(SeqC[-20:].upper() + SeqD)
            block.append(''.join(k))

    for r in range(0, frag, 1):
        for x in range(r, len(block), frag):
            if r == 0:
                block[x] = block[x] + rrestriction
            elif r == (frag-1):
                block[x] = fassembly[:14] + frestriction + block[x][20:]
            else:
                block[x] = fassembly[:14] + frestriction + block[x][20:] + rrestriction

    for p in block:
        lib_export.append(p)

    print(lib_export)
    print(len(lib_export))

    #Set up sequence name list for export
    seq_id = []
    for r in range(0, len(my_assembly)):
        l = [(list(my_assembly.items())[r])[0]]*(frag)

        for name in l:
            seq_id.append(name)

    # Extend constructs with dummy sequence and add final reverse assembly primer
    for t in range(0, len(lib_export)):
        print(t)
        y = [lib_export[t][i:i + payload] for i in range(0, len(lib_export[t]), payload)]
        x = y[len(y) - 1]
        z = payload - len(x) - 20
        y[len(y) - 1] = x + dummy[:z].upper() + rassembly.upper()

        bar_seq = (list(screened_sub.items())[t+160])[1]
        bar_id = (list(screened_sub.items())[t])[0]
        kk = seq_id[t]

        bar_orderlist[bar_id] = (bar_seq)

        #Add initial restriction enzyme
        for f in range(0, len(y)):
            ll = fpool.upper()[:8] + bar_seq + 'gCAGTG' + y[f] + rpool.upper()

            print(ll)

            export.append(SeqRecord(Seq(ll,IUPAC), id= kk + "_" + bar_id, description=''))

    output_handle = open("testing_06072020.fasta", 'w')
    SeqIO.write(export, output_handle, "fasta")
    output_handle.close()

    for ii, pp in bar_orderlist.items():
        order_barcodes.append(SeqRecord(Seq(pp,IUPAC), id = ii, description=''))

    output_handle = open("testing_06072020+primer1.fasta", 'w')
    SeqIO.write(order_barcodes, output_handle, "fasta")
    output_handle.close()





BuildLibrary('testing_06072020.fasta', 'SK_Barcodes.fas', 6, 101)






